Initial delivery of NEMISIS level 1 processing software

kbromund opened the pull request:
HERMES/NEMISIS Level 1 processing software and sample Level 0 data.  
pytest will successfully process Level 0 to Level 1.
This branch was named ccsdspy to distinguish it from an earlier version that did not use ccsdspy.

Alrobbertz commented on hermes_nemisis/calibration/calibration.py:
```diff
@@ -155,11 +149,31 @@ def parse_l0_sci_packets(data_filename: Path) -> dict:
+    for i in range(n_pkts):
+        CK_A = 0 
+        CK_B = 0
+        for octet_num in range(1115):
```
Is this always going to be the same? Is this 1115 a constant or is it potentially variable? If its a constant, maybe ad a comment on where it comes from.

Alrobbertz replied:
Thank you for adding this comment. This makes much more sense.

Alrobbertz commented on hermes_nemisis/calibration/calibration.py:
```diff
@@ -183,23 +197,434 @@ def l0_sci_data_to_cdf(data: dict, original_filename: Path) -> Path:
-        file_metadata["time"],
-        "l1",
-        f'1.0.{file_metadata["version"]}',
+    cdf_version = f'1.0.{int(file_metadata["version"])}'
```
I believe all the files should already contain an `X.Y.Z` version. I know the test file in this PR does not match this, but will the files we receive from the MOC have the same versioning scheme @dbarrous  @ehsteve ?

Alrobbertz commented on hermes_nemisis/calibration/calibration.py:
```diff
@@ -183,23 +197,434 @@ def l0_sci_data_to_cdf(data: dict, original_filename: Path) -> Path:
+    )
+    log.info(f"Finished Creating HermesData structure")
+
+    cdf_file_path = nemisis_data.save(output_path=hermes_nemisis._data_directory, overwrite=True)
```
@dbarrous  is this the correct directory to save to the S3 buckets?

Alrobbertz commented on hermes_nemisis/calibration/readbin_CCSDS.py:
It doesn't look like this file or function is used in the `calibration.py` workflows and it looks like the unit tests for this file are commented out? Is this really needed as a part of this PR? It looks like this is more of a test script and the contents of this function are broken up among other functions in the `calibration.py` file.

Alrobbertz commented on hermes_nemisis/tests/test_calibration.py:
```diff
@@ -24,13 +28,22 @@ def level1_file(tmp_path_factory):
         pass
     return fn
 
+# def test_readbin_CCSDS():  
+#     output_file = readbin_CCSDS('hermes_nemisis/data/hermes_NEM_l0_2024085-174022_v00.bin')
+#     # assert output_file.is_file()  # this fails.  output_file is a string object, not a file. is this a HermesData issue?
```
If the `readbin_CCSDS.py` file is not needed, can you remove these commented out test cases? If it is needed, then you should un-comment the test case and validate that they pass.

kbromund commented:
Hi Andrew,
Please see my responses to each of your comments in line, below.
I have pushed up the corresponding changes to my ccsdspy branch on GitHub
-Ken

From: Andrew Robbertz ***@***.***>
Date: Wednesday, August 21, 2024 at 9:31 AM
To: HERMES-SOC/hermes_nemisis ***@***.***>
Cc: Bromund, Kenneth R. (GSFC-5860) ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] [BULK] Re: [HERMES-SOC/hermes_nemisis] Initial delivery of NEMISIS level 1 processing software (PR #26)
CAUTION: This email originated from outside of NASA.  Please take care when clicking links or opening attachments.  Use the "Report Message" button to report suspicious messages to the NASA SOC.



@Alrobbertz commented on this pull request.

________________________________

In hermes_nemisis/calibration/calibration.py<https://github.com/HERMES-SOC/hermes_nemisis/pull/26#discussion_r1725041388>:

>      data = pkt.load(data_filename)

+

+    # calculate checksum of each packet

+    n_pkts = len(data['CHECKSUM'])

+    packet_bytes = ccsdspy.utils.split_packet_bytes(

+        data_filename,

+        include_primary_header=False

+    )

+    calculated_checksum_out = np.zeros(n_pkts, dtype=np.uint16)

+    for i in range(n_pkts):

+        CK_A = 0

+        CK_B = 0

+        for octet_num in range(1115):

Is this always going to be the same? Is this 1115 a constant or is it potentially variable? If its a constant, maybe ad a comment on where it comes from.



Yes, it will always be the same.  I added a comment.

________________________________

In hermes_nemisis/calibration/calibration.py<https://github.com/HERMES-SOC/hermes_nemisis/pull/26#discussion_r1725045707>:

>      file_metadata = parse_science_filename(original_filename.name)



-    cdf_filename = original_filename.parent / create_science_filename(

-        file_metadata["instrument"],

-        file_metadata["time"],

-        "l1",

-        f'1.0.{file_metadata["version"]}',

+    cdf_version = f'1.0.{int(file_metadata["version"])}'

I believe all the files should already contain an X.Y.Z version. I know the test file in this PR does not match this, but will the files we receive from the MOC have the same versioning scheme @dbarrous<https://github.com/dbarrous> @ehsteve<https://github.com/ehsteve> ?

    In the example L0 files, there is a single, two-digit version number.   I assume that this convention is defined by the MOC, and I think it is fine.   I am adopting a convention where the Z-version of the Level 1 file will reflect the version of the Level 0 file, but without the leading zero (hence the conversion to int before formatting as a string in my code).  I have added a comment to the code to explain this.

We discussed that the instrument data processing code should be responsible for defining the X.Y.Z version for its output files.   The X-version number will be hard-coded, as it is a function of the instrument software version.  The Y-version number will be taken from the calibration file, but this doesn’t apply to Level 1 files, so here, the Y version is hard-coded as 0.   I hope that we can adopt HERMES SOC-wide guidelines for the X.Y.Z version number that are compatible with my intended scheme, but I am open to further discussion.  Note that in this scheme, the X and Y versions are not relevant to Level 0 files: the X.Y.Z scheme need not apply to MOC-generated files.

________________________________

In hermes_nemisis/calibration/calibration.py<https://github.com/HERMES-SOC/hermes_nemisis/pull/26#discussion_r1725052333>:

>

-    return cdf_filename

+    nemisis_data.timeseries["Epoch_pkt"]["hermes_nem_checksum_valid"].meta.update(

+        OrderedDict({"CATDESC":"telemetry packet checksum valid"})

+    )

+    nemisis_data.timeseries["Epoch_pkt"]["hermes_nem_checksum"].meta.update(

+        OrderedDict({"CATDESC":"telemetry packet checksum"})

+    )

+    nemisis_data.timeseries["Epoch_pkt"]["hermes_nem_calculated_checksum"].meta.update(

+        OrderedDict({"CATDESC":"telemetry packet calculated checksum"})

+    )

+    log.info(f"Finished Creating HermesData structure")

+

+    cdf_file_path = nemisis_data.save(output_path=hermes_nemisis._data_directory, overwrite=True)

@dbarrous<https://github.com/dbarrous> is this the correct directory to save to the S3 buckets?

                    Let me know the correct method, and I will adjust the code, as necessary.

________________________________

On hermes_nemisis/calibration/readbin_CCSDS.py<https://github.com/HERMES-SOC/hermes_nemisis/pull/26#discussion_r1725058557>:

It doesn't look like this file or function is used in the calibration.py workflows and it looks like the unit tests for this file are commented out? Is this really needed as a part of this PR? It looks like this is more of a test script and the contents of this function are broken up among other functions in the calibration.py file.

                    Deleted.

________________________________

In hermes_nemisis/tests/test_calibration.py<https://github.com/HERMES-SOC/hermes_nemisis/pull/26#discussion_r1725060893>:

> +# def test_readbin_CCSDS():

+#     output_file = readbin_CCSDS('hermes_nemisis/data/hermes_NEM_l0_2024085-174022_v00.bin')

+#     # assert output_file.is_file()  # this fails.  output_file is a string object, not a file. is this a HermesData issue?

If the readbin_CCSDS.py file is not needed, can you remove these commented out test cases? If it is needed, then you should un-comment the test case and validate that they pass.

                    Done.  I also added some clarifying comments regarding the commented-out test file that is useful to us, but not included in the repo.

—
Reply to this email directly, view it on GitHub<https://github.com/HERMES-SOC/hermes_nemisis/pull/26#pullrequestreview-2250786524>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BBMIDVKGJB63K7SCQURKEK3ZSSJDJAVCNFSM6AAAAABMM6TCBCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDENJQG44DMNJSGQ>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Alrobbertz commented:
Hi @kbromund, thank you for addressing many of the comments here. It does not look like the changes you mentioned were successfully pushed to this branch. Could you try to make sure the changes are committed and pushed up? If you'd like help or a second set of eyes, please send me an email or a message on slack and I'd be happy to take a look.
